File: packages/sdk-python/veltrix_sdk/resources/tailscale_config.py
# -*- coding: utf-8 -*-

from .base import BaseResource


class TailscaleConfigResource(BaseResource):
    """Handles the tenant Tailscale configuration (``/api/tailscale-config``)."""
    RESOURCE_PATH = "tailscale-config"  # Note the hyphen

    def get(self, **kwargs):
        """Retrieves the Tailscale configuration (GET /api/tailscale-config)."""
        return self._list(**kwargs)

    def set(self, tailnet, api_key, api_url=None, **kwargs):
        """Creates/updates the Tailscale configuration (POST /api/tailscale-config)."""
        data = {"tailnet": tailnet, "apiKey": api_key, "apiUrl": api_url}
        data = {k: v for k, v in data.items() if v is not None}
        return self._create(data=data, **kwargs)

    def delete(self, **kwargs):
        """Deletes the Tailscale configuration (DELETE /api/tailscale-config)."""
        return self._http_client.delete(self._get_path(), **kwargs)

    def check(self, **kwargs):
        """Checks whether Tailscale is configured (GET /api/tailscale-config/check)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/check", **kwargs)

File: packages/sdk-js/src/resources/tailscale_config.ts
import { BaseResource } from './base';
import { AxiosRequestConfig } from 'axios';
import { APIError } from '../errors';

// Define interfaces for Tailscale Config data structures
interface TailscaleConfigDetails {
  id: string; // uuid
  apiUrl: string;
  tailnet: string;
  apiKeyConfigured: boolean; // Indicates if API key is set (key itself not returned)
  createdAt: string; // ISO Date string
  updatedAt: string; // ISO Date string
}

interface SetTailscaleConfigPayload {
  tailnet: string;
  apiKey: string;
  apiUrl?: string;
}

interface CheckTailscaleConfigResponse {
    configured: boolean;
    apiUrl?: string | null;
    tailnet?: string | null;
}

export class TailscaleConfigResource extends BaseResource {
  protected readonly RESOURCE_PATH = "tailscale-config"; // Note hyphen

  async get(config?: AxiosRequestConfig): Promise<TailscaleConfigDetails> {
    // Corresponds to GET /api/tailscale-config
    return this._list<TailscaleConfigDetails>(undefined, config); // Use _list for base path GET
  }

  async set(payload: SetTailscaleConfigPayload, config?: AxiosRequestConfig): Promise<TailscaleConfigDetails> {
    // Corresponds to POST /api/tailscale-config
     const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this._create<TailscaleConfigDetails>(cleanedPayload, undefined, config); // Use _create for base path POST
  }

  async delete(config?: AxiosRequestConfig): Promise<{ message: string }> {
    // Corresponds to DELETE /api/tailscale-config
    // API Spec indicates 200 OK with message
    const path = this._getPath();
    // Use http_client directly as base _delete expects ID and handles 204 (null)
    const result = await this.httpClient.delete<{ message: string } | null>(path, config);
    if (result === null) {
        throw new APIError(`DELETE ${path} returned null/204, expected 200 with message.`);
    }
    return result;
  }

  async check(config?: AxiosRequestConfig): Promise<CheckTailscaleConfigResponse> {
    // Corresponds to GET /api/tailscale-config/check
    const path = `${this.RESOURCE_PATH}/check`;
    return this.httpClient.get<CheckTailscaleConfigResponse>(path, config);
  }
}
